Add new `ros2 bag` Kilted features + updates

chen-harrison opened the pull request:
## Description

- Demonstrate the new `ros2 bag play` progress bar
- Provide instructions on how to use new `-i` argument to play multiple bags
- Add an example of recording multiple bags earlier in the tutorial, so as to have an example to use with `-i` explanation
- Add `--topics` to `ros2 bag record` commands to reflect the deprecation of positional topic arguments
- Correct wording so that document is consistent with explaining that a new bag **directory**, not file, is made (with the given `-o` argument, if provided), and pathing to .mcap files now reflects that
- Update output of commands like `ros2 bag record` and `ros2 bag play`, so as not to confuse the reader when their output differs

Fixes #5550

chen-harrison committed:
Add --topics to ros2 bag record Commands

- These changes are meant to reflect the deprecation of positional
  topic arguments in favor of using --topics

chen-harrison committed:
Change from Bag Files to Bag Directories

- The bag directories are correctly explained in 3.1, but in subsequent
  sections it falls back to referring to the recordings as files.
  Adjust wording to explain that a bag directory is made
- Change commands to capture the .mcap file inside the bag directory

chen-harrison committed:
Update ros2 bag Command Output

- Add action and service info for ros2 bag info
- Add playback controls and playback progress for ros2 bag play

chen-harrison committed:
Add Multiple Bag Playback Explanation

- To make this progression logical, add an extra section where a
  recording is split into multiple bag files with -d arg in a bag
  directory called subset_separate
- Later, use these separate files when explaining how to play all bag
  files from a single recording (just call ros2 bag play on the bag
  directory), as well as how to play multiple select files (using new
  arg -i!)

fujitatomoya commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -163,32 +163,58 @@ This directory will contain a ``metadata.yaml`` along with the bag file in the r
 ~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-You can also record multiple topics, as well as change the name of the file ``ros2 bag`` saves to.
+You can also record multiple topics, as well as change the name of the bag directory ``ros2 bag`` saves to.
```
the previous `file` is not actually wrong, it saves the bag file with the specified name as well as bag directory. how about the following?

```suggestion
You can also record multiple topics, as well as change the name of the destination ``ros2 bag`` saves to.
```

christophebedard replied:
I think what matters here is the _directory_. The files themselves are just a component of the bag; their name doesn't really matter. So I personally think we should keep "bag directory"

chen-harrison replied:
I see that the help documentation for `ros2 bag play` says "destination", and maybe that helps clarify that it can also be a path? Would "bag directory path" accomplish that?

Also, since I'm still relatively new as a ROS contributor, I'm not really sure who to listen to as final say :sweat_smile:

christophebedard replied:
[`ros2 bag play`](https://github.com/ros2/rosbag2?tab=readme-ov-file#play) can take a directory (with a `metadata.yaml` file and some storage files, e.g., `.mcap` files) _or_ take a single storage file (e.g., `.mcap`).

However, [`ros2 bag record` takes a name that is then used to create a directory](https://github.com/ros2/rosbag2?tab=readme-ov-file#recording-data-). That section of the Rosbag2 README specifically says "will create a new _folder_".

I think "as well as change the name of the destination" isn't correct. If we use "destination" I think we should remove "name of" and just use "as well as change the destination". Since this is a beginner tutorial, I also think we should keep it simple, and "directory" is simpler/clearer than "destination" for a new user.

Therefore, I would use "bag directory" like you did. @fujitatomoya what do you think?

fujitatomoya commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -163,32 +163,58 @@ This directory will contain a ``metadata.yaml`` along with the bag file in the r
 
-The ``-o`` option allows you to choose a unique name for your bag file.
-The following string, in this case ``subset``, is the file name.
+The ``-o`` option allows you to choose a unique name for your bag directory.
+The following string, in this case ``subset``, is the bag directory name.
```
```suggestion
The ``-o`` option allows you to choose the destination for your bag directory and bag file name.
The following string, in this case ``subset``, is the bag directory name and bag file prefix name.
```

christophebedard replied:
Same comment as above: https://github.com/ros2/ros2_documentation/pull/5771/files#r2211670982

chen-harrison replied:
At the end of section 3.1 when it says 

> This directory will contain a `metadata.yaml`  along with the bag file in the recorded format.

We could replace it with

> This directory will contain a bag file of the same name, with an index suffix "_0". It also contains a `metadata.yaml`  file.

This would establish early on that the files use the name of the bag directory.

christophebedard replied:
I personally don't think we should talk about the file (`$name/$name_0.mcap`) or the prefix of the bag file in the bag directory. This is a beginner tutorial. The mention of "bag file" and `metadata.yaml` at the end of section 3.1 ("This directory will contain a ``metadata.yaml`` along with the bag file in the recorded format.") is enough I think.

fujitatomoya commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -163,32 +163,58 @@ This directory will contain a ``metadata.yaml`` along with the bag file in the r
+    [INFO] [rosbag2_cpp]: Writing remaining messages from cache to the bag. It may take a while
+    [INFO] [rosbag2_cpp]: Writing remaining messages from cache to the bag. It may take a while
+
+Press ``Ctrl+C`` when you're finished.
```
```suggestion
Press :kbd:`Ctrl-C` when you're finished.
```

chen-harrison replied:
Didn't know about this formatting option! Applied it to all other applicable areas of the document.

fujitatomoya reviewed (commented):
overall lgtm with a few minor comments.

@christophebedard @MichaelOrlov either of you can review this?

chen-harrison committed:
``Ctrl+C`` -> :kbd:`Ctrl-C`

- Use keyboard syntax where appropriate

christophebedard commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -198,12 +224,12 @@ You can see details about your recording by running:
-    $ ros2 bag info subset
-    Files:             subset.mcap
+    $ ros2 bag info subset/subset_0.mcap
+    Files:             subset_0.mcap
```
I think we should keep `subset` here for simplicity.

christophebedard commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -212,19 +238,41 @@ Running this command on the ``subset`` bag file will return a list of informatio
 
-    $ ros2 bag play subset
-    [INFO] [rosbag2_storage]: Opened database 'subset'.
+    $ ros2 bag play subset/subset_0.mcap
```
Similar to my other comment, I think we should just keep `subset` here. The user just wants to "play the bag" and not really "play the bag file"

christophebedard commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -212,19 +238,41 @@ Running this command on the ``subset`` bag file will return a list of informatio
 
-Before replaying the bag file, enter ``Ctrl+C`` in the terminal where the teleop is running.
+5.1 Play a single bag file
+~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
5.1 Play a single bag
~~~~~~~~~~~~~~~~~~~~~
```

christophebedard commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -243,6 +291,29 @@ To get an idea of how often position data is published, you can run the command:
     $ ros2 topic hz /turtle1/pose
 
+5.2 Play multiple bag files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
5.2 Play multiple bags
~~~~~~~~~~~~~~~~~~~~~~
```

christophebedard commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -243,6 +291,29 @@ To get an idea of how often position data is published, you can run the command:
+5.2 Play multiple bag files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you want to play multiple bag files that are all a part of the same recording,
+simply call ``ros2 bag play`` on the bag directory, instead of a file:
+
+.. code-block:: console
+
+    $ ros2 bag play subset_separate
+
+This will play all the files inside of the ``subset_separate`` recording consecutively.
+
+If you want to play a specific selection of files, that can be done by calling ``-i <bag_file_name>``
+for each file.
+
+.. code-block:: console
+
+    $ ros2 bag play -i subset_separate/subset_separate_0.mcap -i subset_separate/subset_separate_2.mcap
+
+This will play the first and third bag files in ``subset_separate``, skipping the second.
+They will play with the same relative time as when recorded, meaning there will be a 5 second pause
+in between, where the ``subset_separate_1.mcap`` would have played if not omitted.
```
I think this should be reframed to not mention "the bag _files_ inside the same recording." It should be about "playing multiple bags."

Your explanation here about "They will play with the same relative time as when recorded, meaning there will be a 5 second pause" is correct, but I don't think this is the goal of this feature.

The feature is about playing multiple _bags_ that contain different topics recorded at the same time. This is intended for cases where a lot of data needs to be recorded at the same time, so multiple recorders run at the same time: each one records a subset of the topics. Individual `.mcap` files part of the same bag don't overlap in time (they're sequential), so they're not really separate bags.

This section should also mention how messages from separate bags are ordered when replayed. This is controlled by the `--message-order` option. By default, messages are ordered based on when they were received by the recorder node (which isn't quite the same as when they were recorded):

```console
$ ros2 bag play --help
...
  --message-order {received,sent}
                        The reference to use for bag message chronological ordering. Choices: reception timestamp, publication timestamp. Default: reception timestamp. If messages are
                        significantly disordered (within a single bag or across multiple bags), replayed messages may not be correctly ordered. A possible solution could be to increase
                        the read_ahead_queue_size value to buffer (and order) more messages.
...
```

So I think it should be something like this:

```suggestion
Multiple bags may be recorded at the same time with different topics, for example to distribute the recording workload over multiple recorder processes.
To play multiple bags at the same time, simply call ``ros2 bag play`` and provide bags with ``-i``:

.. code-block:: console

    $ ros2 bag play -i subset -i subset_separate

By default, this will play messages from both bags based on when they were originally received by the recorder.
The chronological ordering of messages can be changed with ``--message-order``.
```

chen-harrison replied:
I see, that makes a lot more sense - I didn't fully understand the original motivations behind the implementation. I originally added the instructions for the `subset_separate` recording to generate multiple bags, which we'd then be able to use for this example.

Given that the best way to demonstrate its use case is two bags recorded simultaneously with different topics, should that `subset_separate` example change? Is it not "beginner enough" to start `ros2 bag record -o vel --topics /turtle1/cmd_vel` and `ros2 bag record -o pose --topics /turtle1/pose` at the same time? Or is it more helpful to learn how to split up bag files, and less so to see the bags playing simultaneously?

christophebedard replied:
> I see, that makes a lot more sense - I didn't fully understand the original motivations behind the implementation. I originally added the instructions for the `subset_separate` recording to generate multiple bags, which we'd then be able to use for this example.

Yeah, I would very intentionally keep these two separate, see below.

> Given that the best way to demonstrate its use case is two bags recorded simultaneously with different topics, should that `subset_separate` example change? Is it not "beginner enough" to start `ros2 bag record -o vel --topics /turtle1/cmd_vel` and `ros2 bag record -o pose --topics /turtle1/pose` at the same time? Or is it more helpful to learn how to split up bag files, and less so to see the bags playing simultaneously?

I think these are two separate features: (1) playing multiple _overlapping_ bags and (2) splitting a _single_ bag into multiple files. Splitting a bag into multiple files can be useful to avoid having huuuuuge files, see https://github.com/ros2/rosbag2?tab=readme-ov-file#splitting-files-during-recording.

For (1), I agree with your simple example: `ros2 bag record -o vel --topics /turtle1/cmd_vel` + `ros2 bag record -o pose --topics /turtle1/pose`. You can show this as an example before showing how to play the two bags (`ros2 bag play -i vel -i pose`) in the same (sub)section.

Separately from that, I would also document (2), but keep it totally separate from (1). And maybe use `-o subset_split` instead of `-o subset_separate` just to use the same terminology, i.e., split(ting)?

For each of these, I would make sure to briefly mention why the feature is useful: (1) replaying overlapping bags that were recorded separately to distribute the recording workload and (2) keeping file sizes under a certain size.

chen-harrison replied:
That all makes sense to me - will start on those changes :+1:

christophebedard reviewed (changes requested):
Thanks for the PR! :rocket: 

I want to clarify some sections to better reflect the feature and the goal behind them, but otherwise this looks good!

chen-harrison committed:
Focus on Bag Recording Instead of File

- For examples that involve calling ros2 bag info and ros2 bag play,
  have the example show it with playing the whole recording instead of
  individual files

chen-harrison committed:
Correct Wording Around ros2 bag info Section

chen-harrison committed:
subset_separate -> subset_split

- Change wording around splitting recording
- Explain the use case of splitting recording

chen-harrison committed:
Overhaul 'Play Multiple Bags' Section

- Add example where there are two separate record commands, each
  recording a topic
- Explain the use case for this separate recording
- Add explanation of --message-order

chen-harrison commented:
@christophebedard I'm pretty new to open-source still so not sure what the cadence of these things are, but letting you know these changes should be ready for review, in case you were waiting for me to notify you.

My only minor concern is that the "play multiple bags" section is a bit lengthy now based on the structure that we discussed, but let me know your thoughts. Thanks!

christophebedard commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -243,6 +293,38 @@ To get an idea of how often position data is published, you can run the command:
+~~~~~~~~~~~~~~~~~~~~~~
+
+At times, it is relevant to split the desired recorded topics amongst multiple recordings, as a way to distribute the recording workload.
+As an example, we can record ``/turtle1/cmd_vel`` and ``/turtle1/pose`` each to their own topic.
```
```suggestion
As an example, we can record ``/turtle1/cmd_vel`` and ``/turtle1/pose`` each to their own bag.
```

christophebedard commented:
> I'm pretty new to open-source still so not sure what the cadence of these things are, but letting you know these changes should be ready for review, in case you were waiting for me to notify you.

Thank you! We're pretty busy, so a simple ping when the PR is ready for review again is helpful :)

> My only minor concern is that the "play multiple bags" section is a bit lengthy now based on the structure that we discussed, but let me know your thoughts. Thanks!

I think it's good, and this way it's pretty self-contained.

christophebedard commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -155,55 +155,83 @@ The movements don't matter, but try to make a recognizable pattern to see when y
+3.3 Split recording into multiple files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can also split your recording into multiple files, based on either recording duration or file size.
+``-d <max_bag_duration>`` ensures that each file only lasts ``<max_bag_duration>`` seconds before it starts writing to
+a new file, or ``-b <max_bag_size>`` ensures that each file does not exceed ``<max_bag_size>`` bytes in file size.
+This prevents large and unwieldy file sizes, and protects against losing all data if the recording operation becomes
+corrupted at some point.
```
```suggestion
You can also split your recording into multiple files, based on either recording duration or file size.
``-d <max_bag_duration>`` ensures that each file only lasts ``<max_bag_duration>`` seconds before it starts writing to a new file, or ``-b <max_bag_size>`` ensures that each file does not exceed ``<max_bag_size>`` bytes in file size.
This prevents large and unwieldy file sizes, and protects against losing all data if the recording operation becomes corrupted at some point.
```

christophebedard commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -155,55 +155,83 @@ The movements don't matter, but try to make a recognizable pattern to see when y
+    [INFO] [rosbag2_cpp]: Writing remaining messages from cache to the bag. It may take a while
+
+Press :kbd:`Ctrl-C` when you're finished.
+You should find a ``subset_split`` directory with these files inside:
+``subset_split_0.mcap``, ``subset_split_1.mcap``, and so on.
```
```suggestion
You should find a ``subset_split`` directory with these files inside: ``subset_split_0.mcap``, ``subset_split_1.mcap``, and so on.
```

christophebedard commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -212,19 +240,41 @@ Running this command on the ``subset`` bag file will return a list of informatio
+    Actions:           0
+    Action information:
+
+Alternatively, you can also call ``ros2 bag info`` on an individual file, such as ``subset_split/subset_split_0.mcap``,
+and it will only show information for that portion of the recording; in this case, the first 5 seconds.
```
```suggestion
Alternatively, you can also call ``ros2 bag info`` on an individual file, such as ``subset_split/subset_split_0.mcap``, and it will only show information for that portion of the recording; in this case, the first 5 seconds.
```

christophebedard commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -243,6 +293,38 @@ To get an idea of how often position data is published, you can run the command:
+At times, it is relevant to split the desired recorded topics amongst multiple recordings, as a way to distribute the recording workload.
+As an example, we can record ``/turtle1/cmd_vel`` and ``/turtle1/pose`` each to their own topic.
+
+Create two terminal instances. In the first one, run the following:
```
```suggestion
Create two terminal instances.
In the first one, run the following:
```

christophebedard commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -243,6 +293,38 @@ To get an idea of how often position data is published, you can run the command:
+
+Move the turtle around as you did before, then end both recordings with :kbd:`Ctrl-C` when finished.
+
+To have these two recordings play in parallel with correct timing, call ``ros2 bag play`` with ``-i <bag_name>``
+for each bag you want to include. In this case, run:
```
```suggestion
To have these two recordings play in parallel with correct timing, call ``ros2 bag play`` with ``-i <bag_name>`` for each bag you want to include.
In this case, run:
```

christophebedard reviewed (commented):
You'll need to make sure that you have only 1 sentence per line (see CI): https://docs.ros.org/en/rolling/The-ROS2-Project/Contributing/Contributing-To-ROS-2-Documentation.html#writing-pages

Other than that and a small fix, this looks good to me!

chen-harrison committed:
One Sentence per Line + Small Typo Fix

- Add/remove line breaks so that each sentence lives on its own line,
  adhering to the CI guidelines

christophebedard commented on source/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.rst:
```diff
@@ -243,6 +289,37 @@ To get an idea of how often position data is published, you can run the command:
+    $ ros2 bag play -i subset_cmd_vel -i subset_pose
+
+This will play the ``subset_cmd_vel`` and ``subset_pose`` recordings together, with the playback synced to replicate the original order of messages.
+If used, the optional argument ``--message-order {received,sent}`` determines whether the messages are sequenced according to the time they were received or published (defaults to received). This applies to playing a single bag as well.
```
Last one: https://github.com/ros2/ros2_documentation/actions/runs/17224339117/job/49124370057?pr=5771

```suggestion
If used, the optional argument ``--message-order {received,sent}`` determines whether the messages are sequenced according to the time they were received or published (defaults to received).
This applies to playing a single bag as well.
```

FYI you can run the documentation tests locally: https://docs.ros.org/en/rolling/The-ROS2-Project/Contributing/Contributing-To-ROS-2-Documentation.html#checking-testing-the-site

chen-harrison replied:
Ah that's my fault. I ran `make lint` locally just now and it didn't find anything else (`make test` is erroring out, need to look into that).

christophebedard replied:
Let me know if I can help!

christophebedard reviewed (approved):
Thanks a lot for the PR and for iterating!
